### PR TITLE
BLE sensors data through a connect, LYWSD03MMC and MHO-C401 add

### PR DESCRIPTION
**Pull request by 1technophile**
Add LYWSD03MMC low-cost BLE temperature sensor
Add the possibility to have 1 BLE connect operation every X BLE scans
Add MHO-C401

### CODE REVIEW

File: docs/use/ble.md [hunk 1]
```diff
@@ -73,7 +73,7 @@ So as to keep your white/black list persistent you can publish it with the retai
 `mosquitto_pub -r -t home/OpenMQTTGateway/commands/MQTTtoBT/config -m '{"white-list":["01:23:14:55:16:15","4C:65:77:88:9C:79","4C:65:A6:66:3C:79"]}'`
 :::
 
-## Setting the time between scans and force a scan
+## Setting the time between BLE scans and force a scan
 
 If you want to change the time between readings you can change the interval by MQTT, if you want the BLE scan every 66seconds:
```
Comment (jmw6773) [inline docs/use/ble.md:78]: Splitting these sentences and adding a space helps the readability.

```
If you want to change the time between readings you can change the interval by MQTT.
For example, if you want the BLE to scan every 66 seconds:
```

File: main/main.ino [hunk 2]
```diff
@@ -1293,6 +1293,7 @@ void stateMeasures() {
   SYSdata["lowpowermode"] = (int)low_power_mode;
 #    endif
   SYSdata["interval"] = BLEinterval;
+  SYSdata["scanbcnct"] = BLEscanBeforeConnect;
```
[the remark this answers is not in the document]
Comment (Legion2) [reply main/main.ino:1296]: We should also think about a refactoring how gateways are integrated in the main.h. Each gateway could be implemented as a class which inherit from a Gateway base class. Then only instances of these classes must be handled in the main.h and we have less `#ifdef ZgatewayXY` in the code.

File: main/ZgatewayBT.ino [hunk 3]
```diff
@@ -566,8 +706,9 @@ unsigned long timebt = 0;
 struct decompose d[6] = {{0, 12, true}, {12, 2, false}, {14, 2, false}, {16, 2, false}, {28, 4, true}, {32, 60, false}};
 
 void setupBT() {
-  BLEinterval = TimeBtw_Read;
+  BLEinterval = TimeBtwRead;
```
Comment (Legion2) [inline main/ZgatewayBT.ino:709]: Remove this here, it is already set in `config_BT.h`

File: main/config_BT.h [hunk 4]
```diff
@@ -29,27 +29,35 @@
...
-#define ServicedataMinLength 29
 
-#ifndef TimeBtw_Read
-#  define TimeBtw_Read 55555 //define default time between 2 scans
+#ifndef Scan_duration
+#  define Scan_duration 10000 //define the time for a scan --WARNING-- changing this value can lead to instability on ESP32
+#endif
+#ifndef ScanPerConnect
+#  define ScanBeforeConnect 10 //define number of scans before connecting to BLE devices (ESP32 only)
+#endif
```
Comment (jmw6773) [inline main/config_BT.h]: Should this be:
```suggestion
#ifndef ScanBeforeConnect
#  define ScanBeforeConnect 10 //define number of scans before connecting to BLE devices (ESP32 only)
#endif
```

**Comment by jmw6773** [comment]
This PR also looks to address issue #725

**Comment by 1technophile** [comment]
> This PR also looks to address issue #725

You are right

File: main/main.ino (hunk 2 above)
[the remark this answers is not in the document]
Comment (1technophile) [reply main/main.ino:1296]: > For the future refactoring this should be done with a stateMeasures callback function in each gateway. Then we don't have to put the gateway related variables in public visibility.

[the remark this answers is not in the document]
Comment (1technophile) [reply main/main.ino:1296]: > We should also think about a refactoring how gateways are integrated in the main.h. Each gateway could be implemented as a class which inherit from a Gateway base class. Then only instances of these classes must be handled in the main.h and we have less #ifdef ZgatewayXY in the code.

I'm really interested in discussing the "ideal" architecture and responsibilities separation. I will begin a thread into the community forum.

**Comment by 1technophile** [comment]
> I'm sorry that I don't have either of these devices to test in my OMG environment.

We all contribute with the means and time we have :-) and I have no problem with that.

File: main/ZgatewayBT.ino [hunk 5]
```diff
@@ -471,9 +605,12 @@ void coreTask(void* pvParameters) {
         if (low_power_mode == 2)
           digitalWrite(LOW_POWER_LED, 1 - LOW_POWER_LED_OFF);
         BLEscan();
-        launch_discovery();
+        // Launching a connect every BLEscanBeforeConnect
+        if (!((millis() / (BLEinterval + Scan_duration)) % BLEscanBeforeConnect))
+          BLEconnect();
```
[the remark this answers is not in the document]
Comment (1technophile) [reply main/ZgatewayBT.ino:610]: I'm going to replace this by a BLEscan counter it will be more reliable.

File: main/config_BT.h (hunk 4 above)
Comment (1technophile) [reply main/config_BT.h]: yes, it is corrected

File: main/ZgatewayBT.ino (hunk 5 above)
[the remark this answers is not in the document]
Comment (1technophile) [reply main/ZgatewayBT.ino]: done

**Comment by jmw6773** [comment]
Should the comment for this variable read something like: `//Number of BLE scans between connection cycles`

https://github.com/1technophile/OpenMQTTGateway/blob/23cbe1f5c2d0b283399505d59cff0ca681e44f9f/main/config_BT.h#L60

**Comment by 1technophile** [comment]
> Should the comment for this variable read something like: //Number of BLE scans between connection cycles

Yes, corrected

File: main/ZgatewayBT.ino [hunk 6]
```diff
@@ -75,6 +88,10 @@ static bool oneWhite = false;
 
 int minRssi = abs(MinimumRSSI); //minimum rssi value
 
+unsigned int scanCount = 0;
```
Comment (Legion2) [inline main/ZgatewayBT.ino:91]: move this into `coreTask`

File: main/ZgatewayBT.ino [hunk 7]
```diff
@@ -611,6 +754,8 @@ bool BTtoMQTT() {
     if (returnedString.equals(F(BLEEndOfDiscovery))) //OK+DISCE
     {
       returnedString.remove(0); //clear data string
+      scanCount++;
+      Log.notice(F("Scan number %d end " CR), scanCount);
```
Comment (Legion2) [inline main/ZgatewayBT.ino:758]: why do we increase scanCount here?

File: main/ZgatewayBT.ino [hunk 8]
```diff
@@ -471,9 +608,12 @@ void coreTask(void* pvParameters) {
         if (low_power_mode == 2)
           digitalWrite(LOW_POWER_LED, 1 - LOW_POWER_LED_OFF);
         BLEscan();
-        launch_discovery();
+        // Launching a connect every BLEscanBeforeConnect
+        if (!(scanCount % BLEscanBeforeConnect) || scanCount == 1)
```
Comment (Legion2) [inline main/ZgatewayBT.ino:612]:
```suggestion
        if (scanCount % BLEscanBeforeConnect == 0)
```

File: main/ZgatewayBT.ino [hunk 9]
```diff
@@ -439,11 +482,105 @@ void BLEscan() {
   BLEScan* pBLEScan = BLEDevice::getScan(); //create new scan
   pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks());
   pBLEScan->setActiveScan(true); //active scan uses more power, but get results faster
-  BLEScanResults foundDevices = pBLEScan->start(Scan_duration, false);
-  Log.notice(F("Found %d devices, scan end deinit controller" CR), foundDevices.getCount());
+  BLEScanResults foundDevices = pBLEScan->start(Scan_duration / 1000, false);
+  scanCount++;
+  Log.notice(F("Found %d devices, scan number %d end deinit controller" CR), foundDevices.getCount(), scanCount);
```
Comment (Legion2) [inline main/ZgatewayBT.ino:487]: The counter should be added where it is used.

File: main/ZgatewayBT.ino (hunk 7 above)
Comment (1technophile) [reply main/ZgatewayBT.ino:758]: It is hidden in the commit, this is for the future publication of the scan count to SYStoMQTT

File: main/ZgatewayBT.ino (hunk 8 above)
Comment (1technophile) [reply main/ZgatewayBT.ino:612]: We need to do the BLEconnect after the first scan if we change it like that we will only do it at `scanCount = BLEscanBeforeConnect`
This is required by the low power mode, this mode puts the ESP to sleep after the BLEscan and the BLEconnect operations.

**Comment by 1technophile** [comment]
Thanks for your comments @Legion2 & @jmw6773 !

**Pull request merged**